File: src/main/java/com/acme/GraalPyContextProvider.java
package com.acme;

import org.graalvm.polyglot.Context;
import org.graalvm.python.embedding.GraalPyResources;
import org.graalvm.python.embedding.VirtualFileSystem;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Comparator;

@Configuration
public class GraalPyContextProvider {
    @Bean
    public Context pythonContext() {
        VirtualFileSystem vfs = VirtualFileSystem.newBuilder()
                .resourceLoadingClass(GraalPyContextProvider.class)
                .build();
        try {
            return GraalPyResources.contextBuilder(vfs)
                    .allowNativeAccess(false)
                    .build();
        } catch (Exception ex) {
            throw new RuntimeException("Could not initialize Python context", ex);
        }
    }
}

File: src/main/kotlin/com/acme/GraalPyContextProvider.kt
package com.acme

import jakarta.enterprise.context.ApplicationScoped
import jakarta.enterprise.inject.Produces
import org.graalvm.polyglot.Context
import org.graalvm.python.embedding.GraalPyResources
import org.graalvm.python.embedding.VirtualFileSystem
import java.nio.file.Paths
import java.nio.file.Files

@ApplicationScoped
class GraalPyContextProvider {
    @Produces
    fun pythonContext(): Context {
        val vfs = VirtualFileSystem.newBuilder()
            .resourceLoadingClass(GraalPyContextProvider::class.java)
            .build()
        var tmpdir = Files.createTempDirectory("graalpy-vfs")
        Runtime.getRuntime().addShutdownHook(Thread {
            Files.walk(tmpdir)
                .sorted(Comparator.reverseOrder())
                .forEach { Files.deleteIfExists(it) }
        })
        GraalPyResources.extractVirtualFileSystemResources(vfs, tmpdir)
        return GraalPyResources.contextBuilder(tmpdir)
            .allowNativeAccess(false)
            .build()
    }
}
